text/3888-const-self-fields.md

```diff
@@ -0,0 +1,319 @@
+- Feature Name: `const_self_fields`
+- Start Date: 2025-11-26
+- RFC PR: [rust-lang/rfcs#3888](https://github.com/rust-lang/rfcs/pull/3888)
+- Rust Issue: [rust-lang/rust#3888](https://github.com/rust-lang/rust/issues/3888)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes per-type fields that can be accessed through a value or trait object using new `const self` and `static const self` syntax:
+
+```rust
+impl Foo{
+    const self METADATA_FIELD: i32 = 5;
+    static const self STATIC_METADATA_FIELD: i32 = 10;
+}
+trait Bar {
+    const self METADATA_FIELD: i32;
+    static const self STATIC_METADATA_FIELD: i32;
+}
+```
+This allows code like:
+```rust
+fn use_bar(bar: &dyn Bar) {
+    let x: i32 = bar.METADATA_FIELD; // const self
+    let y: &'static i32 = &bar.STATIC_METADATA_FIELD; // static const self
+}
+fn use_foo(foo: &Foo) {
+    let x: i32 = foo.METADATA_FIELD; // const self
+    let y: &'static i32 = &foo.STATIC_METADATA_FIELD; // static const self
+}
+```
+When combined with traits, enables object-safe, per-implementation constant data that can be read through `&dyn Trait` in a more efficient manner than a dynamic function call, by storing the constant in trait object metadata instead of as a vtable method.
+# Motivation
+[motivation]: #motivation
+Today, Rust has associated constants on types and traits:
+```rust
+trait Foo {
+    const VALUE: i32;
+}
+
+impl Foo for MyType {
+    const VALUE: i32 = 5;
+}
+```
+For monomorphized code where `Self` is known, `MyType::VALUE` is an excellent fit. 
+
+However: You cannot directly read an associated const through a `&dyn Foo`. There is no stable, efficient way to write `foo.VALUE` where `foo: &dyn Foo` and have that dynamically dispatch to the concrete implementation’s const value. 
+
+The common workaround is a vtable method:
+```rust
+trait Foo {
+    fn value(&self) -> i32;
+}
+```
+
+This forces a dynamic function call, which is very slow compared to the `const self` and `static const self` equivalent, and does not have as much compiler optimization potential.
+
+When using a trait object, `const self` and `static const self` store the bits directly inside the vtable, so accessing it is around as performant as accessing a field from a struct, which is of course, much more performant than a dynamic function call.
+
+Imagine a hot loop walking over thousands of `&dyn Behavior` objects every frame to read a tiny “flag”. If that’s a virtual method, you pay a dynamic function call on every object. With `const self` and `static const self`, you’re just doing a metadata load, so the per-object overhead is noticeably much smaller.
+
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### What is const self?
+
+`const self` introduces metadata fields: constants that belong to a type (or trait implementation) but are accessed through a `self` expression. 
+
+Example:
+
+```rust
+struct Foo;
+
+impl Foo {
+    const self CONST_FIELD: u32 = 1;
```

Comment:
I think it’s pointless to allow `const self` fields on inherent impls (except maybe for macros?), so I’d favor only allowing them on trait impls (unless someone can convince me otherwise).

Reply:
For reference, associated types are currently allowed in stable rust only in trait impls.

I don't see any reason to allow `const self` in inherent impls.

Reply:
> For reference, associated types are currently allowed in stable rust only in trait impls.

afaict that's mostly because of a limitation of the trait solver, rather than than Rust deciding we don't want associated types in inherent impls. it is available as a nightly feature, though iirc it's incomplete and buggy.

> I don't see any reason to allow `const self` in inherent impls.

if you want to have things on your struct that act like fields (so you can do `my_struct.some_field`) but aren't actually stored in your struct, this is a good way (though it doesn't match the preferred name casing). abusing `Deref` also works in some cases.